Natspec: Fix internal error when different return name was inherited

Marenz opened the pull request:
fixes #9947

Marenz committed:
Natspec: Fix internal error when different return name was inherited

hrkrshnn commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -2022,6 +2023,49 @@ BOOST_AUTO_TEST_CASE(dev_explicit_inehrit_complex)
+			{
+				"returns":
+				{
+					"_0": "y value"
```
Is this how we want to do it? Maybe we shouldn't inherit the `returns` tag in such cases. If we do, then it would make sense to replace "y" by the name of the current return value.

Marenz replied:
well, that would be the default behavior if there is no name given, it made sense to me to default to that if there is a conflict in the parameter names..
Replacing was also an idea I had, but that's not as easy as it maybe sounds with how the code is currently structured.

Marenz replied:
Oh, I see what you mean. Maybe a better alternative would be to complete skip it then.

hrkrshnn replied:
> Replacing was also an idea I had, but that's not as easy as it maybe sounds with how the code is currently structured.

I agree. The current Natspec code needs an overhaul.

chriseth replied:
I think this line should say
```suggestion
					"z": "value"
```

Marenz replied:
If we only have one parameter, correcting it can be done, but if we have several, how should we proceed? Use the order of the parameters to see which one should be which? Skip the whole thing?

chriseth replied:
There should be a proper correspondence between the parameters, isn't there?

Marenz replied:
if we use the order, yes..

hrkrshnn commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -178,12 +178,15 @@ Json::Value Natspec::extractReturnParameterDocs(std::multimap<std::string, DocTa
 			string content = i->second.content;
 
-			if (paramName.empty())
+			auto nameEndPos = content.find_first_of(" \t");
```
Question: can you give an example of the string content?

Marenz replied:
the content could be `y value`, basically the part after `@return ` and it has to start with the name of the parameter

chriseth commented:
Can you check if we have the same problem with parameters?

Marenz commented:
> Can you check if we have the same problem with parameters?

I checked and they don't have the same problem

Marenz commented:
For normal parameters it seems they are only inherited if they match exactly.

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -48,9 +48,37 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
 
-	for (auto const& [tag, content]: sourceDoc.docTags)
-		if (tag != "inheritdoc" && !existingTags.count(tag))
-			_target.docTags.emplace(tag, content);
```
Do you think it's possible to just do a regex_replace on the content here?

Also not sure why a multimap was chosen for this. Probably a vector of tags would be a better choice here. Tag can be a variant of structs notice, dev, param and return. The return struct can be strings name and description. The number of tags is very small (<20, probably like 5 in actual contracts). Do you think it's worth changing this now?

Marenz replied:
Tags can exist multiple times, like several `@param` tags and such. That's why it's a multimap. If we do such a change I would do it in another PR and not here.. generally I am not opposed to a restructuring of the tags.

hrkrshnn replied:
I'll link https://github.com/ethereum/solidity/issues/8247 for future reference.

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,7 +36,7 @@ using namespace solidity::frontend;
 {
 
-void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDeclaration const*> const& _baseFunctions)
+void copyMissingTags(set<CallableDeclaration const*> const& _baseFunctions, StructurallyDocumentedAnnotation& _target, CallableDeclaration const* _callable = nullptr)
```
Would it make sense to change the first argument from a set to just `CallableDeclaration const* _baseFunction`?

Marenz replied:
From the perspective of the function yes. From the perspective of the caller no, because in 50% of calls I have that parameter only as `set` and would need to do the size check before the call. This way I do the size check only once in the function.

hrkrshnn replied:
True. So maybe not worth the change.

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -48,9 +48,37 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
+			if (_callable && it->first == "return")
+			{
+				size_t docParaNameEndPos = it->second.content.find_first_of(" \t");
+				string const docParameterName = it->second.content.substr(0, docParaNameEndPos);
+
+				if (docParameterName != _callable->returnParameters().at(n)->name())
+				{
+					bool baseHasNoName = (*_baseFunctions.begin())->returnParameters().at(n)->name().empty();
+					newContent.content =
+						_callable->returnParameters().at(n)->name() + (
+							string::npos == docParaNameEndPos || baseHasNoName ?
+							" " + newContent.content :
+							newContent.content.substr(docParaNameEndPos)
+						);
+				}
```
Maybe we can replace this by

```cpp
regex_replace(
     it->second.content,
     regex{"[^\\s\\t]"},
     _callable->returnParameters().at(n)->name(),
     regex_constants::format_first_only
)
```

Marenz replied:
Hmm generally I try to avoid the complexity of regex when I don't really need it.

leonardoalt commented on test/libsolidity/smtCheckerTests/operators/slice_default_end.sol:
```diff
@@ -8,13 +8,12 @@ contract C {
 		assert(bytes(b[10:])[5] == 0xff);
-		assert(bytes(b[10:])[19] == 0xaa);
+		// Removed because of Spacer nondeterminism
+//		assert(bytes(b[10:])[19] == 0xaa);
```
Can you change this to
```suggestion
		//assert(bytes(b[10:])[19] == 0xaa);
```
I grep for `//assert` to find which ones I should try again when there's a new version

Marenz replied:
Done

axic commented:
What is blocking this?

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -48,9 +48,37 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
-	for (auto const& [tag, content]: sourceDoc.docTags)
-		if (tag != "inheritdoc" && !existingTags.count(tag))
-			_target.docTags.emplace(tag, content);
+	for (auto it = sourceDoc.docTags.begin(); it != sourceDoc.docTags.end();)
```
Please either add a comment to explain what this does or extract it into a function which is then self-explanatory through its name.

Marenz replied:
it's used to get information about the return parameters and their names so they can be adjusted when copied in case of functions.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,7 +36,7 @@ using namespace solidity::frontend;
 {
 
-void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDeclaration const*> const& _baseFunctions)
+void copyMissingTags(set<CallableDeclaration const*> const& _baseFunctions, StructurallyDocumentedAnnotation& _target, CallableDeclaration const* _callable = nullptr)
```
Can you find a better name than `_callable`?

chriseth commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -2022,6 +2023,187 @@ BOOST_AUTO_TEST_CASE(dev_explicit_inehrit_complex)
+		}
+
+		contract B is A {
+			function g(int x) public pure override returns (int z, int y) { return (1, 2); }
```
Can you change this to remove `y`?

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,51 @@ using namespace solidity::frontend;
 {
 
-void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDeclaration const*> const& _baseFunctions)
+void copyMissingTags(set<CallableDeclaration const*> const& _baseFunctions, StructurallyDocumentedAnnotation& _target, CallableDeclaration const* _declaration = nullptr)
```
I still don't understand the purpose of the third parameter.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,51 @@ using namespace solidity::frontend;
 	if (_baseFunctions.size() != 1)
 		return;
 
 	auto& sourceDoc = dynamic_cast<StructurallyDocumentedAnnotation const&>((*_baseFunctions.begin())->annotation());
```
Are you only using the first element of a pointer-based set? Can this cause nondeterminism?

Marenz replied:
I am, but it's a set of the size() == 1 as you can see in line 41. 
I am mainly passing a set to avoid at least 4 lines of casting and extracting `_baseFunctions.begin()` before calling the function. I figured I need to do that several times, I might as well do it in the function.

chriseth reviewed (commented):
I'm sorry, but I have a hard time reviewing this. Is it possible to do this without the use of iterators (or at least without using such descriptive names as `it->first` and `it->second`? :) )

Also, where did `existingTags` go or why is it not needed anymore?

Marenz commented:
`existingTags` was no longer needed when I changed to use iterators with the `upper_bound()` function.

Before we had one loop over a multimap. `return` tags for example could have multiple values. `existingTags` was there to make sure we don't copy the already existing ones and we couldn't just check the target map if the tag already exists because due to the multi map we need to add the same tag with different values (e.g. multiple `return`s)

I rewrote it a bit to not use `it->first` and `it->second` much.. I could rewrite it without iterators I suppose but then I'd need to use `equal_range()` at the very least and it will also be not be much easier I fear..

chriseth commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -66,6 +66,7 @@ struct DocTag
 {
 	std::string content;	///< The text content of the tag.
 	std::string paramName;	///< Only used for @param, stores the parameter name.
+	bool inherited = false; ///< True if this tag was inherited
```
How is this used?

Marenz replied:
Ah, I think that was from a different attempt and was kept. Will remove.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,53 @@ using namespace solidity::frontend;
-void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDeclaration const*> const& _baseFunctions)
+void copyMissingTags(set<CallableDeclaration const*> const& _baseFunctions, StructurallyDocumentedAnnotation& _target, CallableDeclaration const* _declaration = nullptr)
 {
 	if (_baseFunctions.size() != 1)
```
Please add a comment: Only copy if there is exactly one direct base function.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,53 @@ using namespace solidity::frontend;
+			DocTag newContent = content;
+			newContent.inherited = true;
+
+			// Update the parameter name for @return tags
```
What about `@param`?

Marenz replied:
well, this issue was about `@return`. As far as I know, `@param` can't be unnamed so they don't have the same problem.
Of course the name can change and in that case nothing will be inherited.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,53 @@ using namespace solidity::frontend;
+					bool baseHasNoName = (*_baseFunctions.begin())->returnParameters().at(n)->name().empty();
+					string const paramName = _declaration->returnParameters().at(n)->name();
+					newContent.content =
+						(paramName.empty() ? "" : std::move(paramName) + " ") + (
```
moving from a const string?

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,53 @@ using namespace solidity::frontend;
+					string const paramName = _declaration->returnParameters().at(n)->name();
+					newContent.content =
+						(paramName.empty() ? "" : std::move(paramName) + " ") + (
+							string::npos == docParaNameEndPos || baseHasNoName ?
```
What if we have `@returns x` (without description)?

Marenz replied:
in that case it is assumed that `x` is the description and the position is used to identify which return parameter it is referring to.

Marenz replied:
in terms of code, `returns` is not part of `content`, so the `docParaNameEndPos` would be `npos` as it couldn't find any whitespace or `\t`.

chriseth replied:
Indeed, but then it would not substitute the name even though it should, right?

Marenz replied:
It wouldn't and I don't think it should either as it isn't detected as "name" but as "description" instead.. either way it wouldn't have much use to have a name without a description OR a description that is just the name...

chriseth replied:
Can you add a test for this case?

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,53 @@ using namespace solidity::frontend;
+						(paramName.empty() ? "" : std::move(paramName) + " ") + (
+							string::npos == docParaNameEndPos || baseHasNoName ?
+							newContent.content :
+							newContent.content.substr(docParaNameEndPos+1)
```
```suggestion
							newContent.content.substr(docParaNameEndPos + 1)
```

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -36,21 +36,53 @@ using namespace solidity::frontend;
+		// Iterate over all values of the current tag (it's a multimap)
+		for (auto next = sourceDoc.docTags.upper_bound(tag); it != next; it++, n++)
+		{
+			DocTag const& content = it->second;
```
Do you really still need the old copy?

Marenz replied:
Yes, that's the documentation for the base function

chriseth replied:
I mean why don't you use `DocTag newContent = content` and replace all mentions of `content` below by `newContent`? (or call everything just `content')
